loadbalancer/algorithm.go

```diff
@@ -220,14 +221,56 @@ func hash(s string) uint32 {
 	return h.Sum32()
 }
 
-// Returns index of endpoint with closest hash to key's hash
-func (ch consistentHash) search(key string) int {
+// returns index in hash ring with the closest hash to key's hash
+func (ch consistentHash) searchRing(key string) int {
 	h := hash(key)
 	i := sort.Search(ch.Len(), func(i int) bool { return ch[i].hash >= h })
 	if i == ch.Len() { // rollover
 		i = 0
 	}
-	return ch[i].index
+
+	return i
+}
+
+// Returns index of endpoint with closest hash to key's hash
+func (ch consistentHash) search(key string) int {
+	ringIndex := ch.searchRing(key)
+	return ch[ringIndex].index
+}
+
+func computeLoadAverage(ctx *routing.LBContext) float64 {
+	sum := 1.0 // add 1 to include the request that just arrived
+	endpoints := ctx.Route.LBEndpoints
+	numEndpoints := len(ctx.Route.LBEndpoints)
+	for _, v := range endpoints {
+		sum += float64(v.Metrics.GetInflightRequests())
+	}
+	if numEndpoints == 0 {
+		return 0.0
+	} else {
+		return sum / float64(numEndpoints)
+	}
+}
+
+// Returns index of endpoint with closest hash to key's hash, which is also below the target load
+func (ch consistentHash) boundedLoadSearch(key string, balanceFactor float64, ctx *routing.LBContext) int {
+	ringIndex := ch.searchRing(key)
+	averageLoad := computeLoadAverage(ctx)
+	targetLoad := averageLoad * balanceFactor
+	// Loop round ring, starting at endpoint with closest hash. Stop when we find one whose load is less than targetLoad.
+	ringIndex -= 1
+	for i := 0; i < ch.Len(); i++ {
+		ringIndex = (ringIndex + 1) % ch.Len()
+		endpointIndex := ch[ringIndex].index
+		inflightRequests := ctx.Route.LBEndpoints[endpointIndex].Metrics.GetInflightRequests()
+		// We know there must be an endpoint whose load <= average load.
+		// Since targetLoad >= average load (balancerFactor >= 1), there must also be an endpoint with load <= targetLoad.
+		if inflightRequests <= int(targetLoad) {
```

Comment:
Maybe cast `inflightRequests` to `float64` instead?

Reply:
```go
ep := ctx.Route.LBEndpoints[endpointIndex]
load := float64(ep.Metrics.GetInflightRequests())
```

Reply:
In the paper it said to pick floor or ceiling of targetLoad so I cast to int to get the floor in this case. I'll switch the naming to load from inflightRequests though